Fix assertion failure that happend when codegen the conversion procedure of converting from calldata array

sifmelcara opened the pull request:
### Description

Fix #6835 

Runtime test work in progress

### Checklist
- [ ] Code compiles correctly
- [ ] All tests are passing
- [ ] New tests have been created which fail without the change (if possible)
- [ ] README / documentation was extended, if necessary
- [ ] Changelog entry (if change is visible to the user)
- [ ] Used meaningful commit messages

sifmelcara commented on libsolidity/codegen/ArrayUtils.cpp:
```diff
@@ -1055,28 +1055,27 @@ void ArrayUtils::accessIndex(ArrayType const& _arrayType, bool _doBoundsCheck, b
 	switch (location)
 	{
 	case DataLocation::Memory:
```
I hope this rewrite will make it easier to read

sifmelcara commented on libsolidity/codegen/CompilerUtils.cpp:
```diff
@@ -923,24 +929,76 @@ void CompilerUtils::convertType(
+						ArrayUtils(m_context).accessIndex(typeOnStack, false, false);
 						StorageItem(m_context, *typeOnStack.baseType()).retrieveValue(SourceLocation(), true);
+						break;
+					case DataLocation::CallData:
```
Copied from ExpressionCompiler's indexAccess implementation

ekpyron replied:
Can you extract it to a public member function of ``CompilerUtils`` then and call it from ``ExpressionCompiler::visit(IndexAccess const&)``?

sifmelcara replied:
Ok

sifmelcara commented:
The bug is: When `convertType()` tries to recursively convert nested calldata array, it forgot to deal with the case that `typeOnStack` is calldata. (Probably because nested calldata array is not a thing before ABIEncoderV2)

ekpyron commented on libsolidity/codegen/ArrayUtils.cpp:
```diff
@@ -1055,28 +1055,27 @@ void ArrayUtils::accessIndex(ArrayType const& _arrayType, bool _doBoundsCheck, b
 	case DataLocation::Memory:
+		// stack: <base_ref> <index>
+		if (!_arrayType.isByteArray())
+			m_context << u256(_arrayType.memoryHeadSize()) << Instruction::MUL;
```
Is this correct for arrays with dynamically encoded base types? E.g.
```
struct S { uint256 a; uint256 b; }
function f(S[] memory s) public {...}
```
It might be that ``memoryHeadSize()`` will be ``64`` in this case, but we want to multiply by ``32`` instead - I'm not sure, but we should check.

ekpyron replied:
Similarly ``function f(uint[2][] memory arr) public {...}``.

ekpyron replied:
*If* this is incorrect, it'd be spooky that we don't have test failures, but we might just not have enough tests for these cases yet...

ekpyron replied:
I'm probably mistaken here and this is fine, but I'll need to recheck.

sifmelcara replied:
Hmm... good catch, I guess it should be `_arrayType.baseType()->memoryHeadSize()` instead of `_arrayType.memoryHeadSize()`, right? 

> It might be that memoryHeadSize() will be 64 in this case

Because it is a reference type, it should be 32
https://github.com/ethereum/solidity/blob/dd04a35c0e2d2111ff165013b0c7fe1a31beb3a7/libsolidity/ast/Types.h#L636

> If this is incorrect, it'd be spooky that we don't have test failures

I think the reason is that currently in solidity codebase, all invocation to `memoryHeadSize()` would returns 32. For types occupy less than 32 bytes we pads, and types occupy more than 32 bytes are all reference type. This is just my guess though, we should check that this is indeed the case.

ekpyron replied:
Ah right - I overlooked the override of ``memoryHeadSize`` in ``ReferenceType`` and only looked at the implementation in ``Type`` that falls back to ``calldataEncodedSize``. You're right: since ``memoryHeadSize()`` always returns ``32`` here, it should be fine in all cases.

ekpyron replied:
Or should it :-)? I'll take some more time looking into it :-).

ekpyron replied:
It still looks somewhat wrong to me and feels like ``_arrayType.baseType()->memoryHeadSize()`` should be correct, if the base type is statically encoded and ``32`` should be correct, if the base type is dynamically encoded - so exactly what the ``CallData`` case does below... but I'm probably missing something myself right now and so far I haven't been able to actually break this...

ekpyron replied:
This shouldn't be an issue for this PR in any case, since it did the same thing before and I'm probably just not seeing why this is correct right now...

ekpyron replied:
Ah ok, now I see - I was thinking about calldata layout the whole time, for which statically sized arrays are encoded in-place, whereas memory arrays always use tail pointers anyways, so this *is* correct - nevermind :-)!

chriseth commented on test/libsolidity/semanticTests/abiEncoderV2/calldata_array_convert_to_memory.sol:
```diff
@@ -0,0 +1,32 @@
+// ====
+// EVMVersion: >homestead
+// ----
+// dummy() -> 42
```
Is this still work in progress?

chriseth commented on libsolidity/codegen/CompilerUtils.cpp:
```diff
@@ -923,24 +929,76 @@ void CompilerUtils::convertType(
 				// stack: <mem start> <source ref> (variably sized) <length> <mem data pos>
 				if (targetType.baseType()->isValueType())
 				{
-					solAssert(typeOnStack.baseType()->isValueType(), "");
```
Can you relax the condition instead of removing the assertion altogether?

sifmelcara commented:
I moved part of this PR's content to #6965 , as the clean up part and refactor part can be reviewed separately

sifmelcara committed:
Fix converting from nested calldata array

It forgots to recursively convert nested calldata array. This bug should
always cause assertion failure instead of generating incorrect bytecode.

sifmelcara committed:
Add tests that assign nested calldata array to nested memory array

sifmelcara commented:
Hmm... Actually this PR only fixes `calldata -> memory` conversion. I just discovered that `calldata -> storage` conversion is also broken. I guess these two issue are independent though, looks like there is something wrong in `ArrayUtils::copyArrayToStorage()`. I will try to find the problem tomorrow or this weekend if I have time.

sifmelcara commented:
Also it is strange that test suite throws exception if I add this function call to the test
```
// f(uint256[][])  : 0x20, 3, 0x60, 0xC0, 0x140, 2, 123, 124, 3, 223, 224, 225, 2, 323, 324 -> 32, 480, 0x20, 3, 0x60, 0xC0, 0x140, 2, 123, 124, 3, 223, 224, 225, 2, 323, 324
```
```
test/boostTest.cpp(125): error: in "semanticTests/abiEncoderV2/storage_array_encoding": Exception during extracted test: Throw location unknown (consider using BOOST_THROW_EXCEPTION)
Dynamic exception type: langutil::Error
std::exception::what: Literal encoding invalid.
[dev::tag_comment*] = Literal encoding invalid.
```
Maybe I missed something obvious?

chriseth commented:
The `0x140` is the problem.

chriseth commented:
Is this ready for review or not?

sifmelcara commented:
Yes

On Mon, Jun 24, 2019, 8:08 PM chriseth <notifications@github.com> wrote:

> Is this ready for review or not?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ethereum/solidity/pull/6881?email_source=notifications&email_token=ACQCRP6C3HL5W6FCKOGGRTDP4C2K3A5CNFSM4HSHCWMKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYMWS6I#issuecomment-504981881>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACQCRP5RSUSTRQMOWQUKAGDP4C2K3ANCNFSM4HSHCWMA>
> .
>

chriseth commented:
Looks good, but here, I would also like @ekpyron to take another look.

ekpyron reviewed (commented):
On a casual vacation-look this looks good to me - but some cases only occur with three-dimensional arrays, did you test that? I think it may be worthwhile to add another test case for that. It should be straightforward to duplicate or extend the ``calldata_array_dynamic_three_dimensional`` test in ``SolidityEndToEndTest.cpp`` to that end (we would basically only need the ``reenc`` part, but with an additional ``calldata``-``memory`` conversion before passing the array on to the ``test`` function).

So probably adding something like
```
				function reenc_memory()" + arrayType + R"( calldata a, uint256 i, uint256 j, uint256 k) external returns (uint256) {
				        " + arrayType + R"( memory b = a;
					return this.test(b, i, j, k);
				}
```
to the contract in the test and calling that just the same way as ``reenc`` is already called should work.

sifmelcara committed:
Allow dynamic calldata struct assignment to memory

sifmelcara commented:
@chriseth could you explain why `0x140` is the problem? It might not be possible to avoid using larger number like `0x140` if we want 3 dimensional array in semantic tests.

@ekpyron What cases only occur with three-dimensional array? (Maybe I overlook some side-effects happend when recursively converting sub-arrays) If you provide an example, I can trace the code to see if it functions correctly before I start writing SolidityEndToEndTests.

sifmelcara commented on libsolidity/codegen/CompilerUtils.cpp:
```diff
@@ -975,9 +993,6 @@ void CompilerUtils::convertType(
 		solAssert(targetTypeCategory == stackTypeCategory, "");
 		auto& targetType = dynamic_cast<StructType const&>(_targetType);
 		auto& typeOnStack = dynamic_cast<StructType const&>(_typeOnStack);
-		solAssert(
```
This assert is redundant. There is already an assertion in the following `switch`

sifmelcara commented on libsolidity/codegen/CompilerUtils.cpp:
```diff
@@ -1030,7 +1045,6 @@ void CompilerUtils::convertType(
 			}
 			case DataLocation::CallData:
 			{
-				solUnimplementedAssert(!typeOnStack.isDynamicallyEncoded(), "");
```
We should let the following `abiDecode()` call decide whether dynamically encoded type is allowed in calldata struct. (If ABIEncoderV2 is enabled, it should be allowed)

ekpyron commented:
@sifmelcara 
* In the test expectations hex literals are parsed as in solidity and require byte-padding, i.e. you need to use ``0x0140`` instead of ``0x140``.
* An example of a case that only occurs in three dimensional arrays is arrays with dynamically encoded, but statically sized base types. I'm not necessarily saying that these cases won't work, but in any case they should be tested.

sifmelcara commented:
@ekpyron 

* Ok I opened #7082 to track the issue.
* Actually those cases indeed won't work in #6983 ... Fortunately you reviewed these PRs.

fubar commented:
@sifmelcara @ekpyron can you give an update on where this is at? We ran into the same issue with multi-dimensional calldata bytes arrays. Thanks!

chriseth commented:
@fubar we are currently not actively working on this because we have to re-check that decoding and length checks for all kinds of calldata variables works properly.

fubar commented:
@chriseth is that work tracked in a github issue or PR? I'd like to be able to follow the progress.

chriseth commented:
@fubar the ones linked here.

Marenz commented:
The linked issue is closed and this needs rebasing. Any other steps required to move this forward?

ekpyron commented:
We decided that we don't want to properly implement this for the old ABIEncoder, but instead want to push ABIEncoderV2 forward, so I'm closing this. I'll open a new issue for improving error messages in these cases.
